---
title: "DESeq"
format:
  html:
    df-print: default
    fig-format: png
    fig-dpi: 300
    execute:
      dev: ragg_png
---

# Load environment and data

```{r}
library(tidyverse)
library(DESeq2)
library(fgsea)
library(msigdbr)
```

##### Check that colnames in count data and row names of metadata are in the same order

```{r}
all(metadata_df_08$sample == colnames(gene_count_08))
```

# DESeq Condition

```{r}
dds_08 <- DESeqDataSetFromMatrix(
  countData = gene_count_08, 
  colData = metadata_df_08,
  design = ~ groups)
```

### PCA plot

converting counts to integer mode

```{r}
dds_08
```

```{r}
pca_08 <- plotPCA(rlog(dds_08), intgroup=c("groups"))
pca_08
```

PCA does not seem to capture healthy/parkinsons. There seem to be other factors which influence the gene expression other than the condition

### Testing for differential expression

```{r}
dds_08 <- DESeq(dds_08)
```

```{r}
res_08 <- results(dds_08)
```

How many genes were significant using an alpha of 0.05?

```{r}
n_sig_08 <- sum(res_08$padj < 0.05, na.rm = TRUE)
n_sig_08
```

```{r}
nrow(gene_count_08)
```

Out of 67,186 genes, 518 showed significant differential expression between the two groups/conditions. We expect a 5% false discovery rate among the genes we call significant.

##### Correcting for gender confounder

Confounders are external variables that are linked to both the independent and dependent variables in a study, distorting the true relationship between them and leading to a misleading conclusion. Let’s check if there is a gender confounder in the data

### MA-plot

```{r}
plotMA(res_08, ylim=c(-2,2))
dev.off()

MA_08 <- plotMA(res_08, ylim=c(-2,2))

```

```{r}
ggsave("../fig/MA_08.png", plot = MA_08)
png(filename = "../fig/MA_08.png")
```

# DESeq Condition&Gender

```{r}
metadata_df_08
```

```{r}
dds_gender_08 <- DESeqDataSetFromMatrix(
  countData = gene_count_08,
  colData = metadata_df_08,
  design = ~ gender + groups)
```

```{r}
dds_gender_08
```

```{r}
pca_08_gender <- plotPCA(rlog(dds_gender_08), intgroup=c("groups", "gender"))
pca_08_gender
```

PC1 does not explicitly seem to capture healthy vs. parkinson's condition.

PC2 also does not seem to separate based on gender.

Or vice versa.

This indicates that gender effects may also contribute to the overall variance in gene expression

```{r}
ggsave("../fig/pca_08_gender.png", plot = pca_08_gender)
```

```{r}
dds_gender_08 <- DESeq(dds_gender_08)
```

A co-factor is *any variable that explains expression variation*, even if it is balanced between conditions and therefore not a confounder.

```{r}
res_gender_08 <- results(dds_gender_08,
                         contrast =c(
                          "groups",
                          "healthy",
                          "parkinson's")
                         )
```

How many genes were significant using an alpha of 0.05?\

```{r}
n_sig_gender_08 <- sum(res_gender_08$padj < 0.05, na.rm = TRUE)
n_sig_gender_08
```

What is the overlap between the differential analysis with and without modeling gender?\

```{r}
signif_08 <- rownames(res_08)[!is.na(res_08$padj) & res_08$padj < 0.05]
signif_gender_08 <- rownames(res_gender_08)[!is.na(res_gender_08$padj) & res_gender_08$padj < 0.05]

# Number of genes in common
n_common_08 <- length(intersect(signif_08, signif_gender_08))
n_common_08
```

Gender looks like a modest but real co-factor.

Without gender in the model we get 518 DE genes; with gender we get 433, with 405 overlapping. That means gender doesn’t dominate the signal but it shiftsmores 0ish genes (a 6,5% reduction).

Even though the PCA does not totally agree by separating, the analysis showed that adjusting for gender helps attribute the condition effect more clearly.

```{r}
saveRDS(res_gender_08, file = "../data/deseq2_results_08.rds")
```

# DESeq Condition&Gender&Age

Correcting for age confounder (Skriv til Kristoffer om PCA. Lav to modeller og beregn signifikante)

```{r}
dds_gender_age_08 <- DESeqDataSetFromMatrix(
  countData = gene_count_08,
  colData = metadata_df_08,
  design = ~  groups + age + gender)
```

```{r}
dds_gender_age_08
dds_gender_age_08 <- DESeq(dds_gender_age_08)
```

```{r}
pca_gender_age_08 <- plotPCA(rlog(dds_gender_age_08), intgroup=c("groups","age", "gender"))
pca_gender_age_08
```

```{r}
ggsave("../fig/pca_gender_age_08.png", plot = pca_gender_age_08)
```

```{r}
res_gender_age_08 <- results(dds_gender_age_08,
                         contrast =c(
                          "groups",
                          "healthy",
                          "parkinson's")
                         )

# How many genes were significant using an alpha of 0.05?
n_sig_gender_age_08 <- sum(res_gender_age_08$padj < 0.05, na.rm = TRUE)
cat("number of significant genes when modelling with gender and age:", n_sig_gender_age_08, "\n")

# What is the overlap between the differential analysis with and without modeling age?
signif_gender_08 <- rownames(res_08)[!is.na(res_gender_08$padj) & res_gender_08$padj < 0.05]
signif_gender_age_08 <- rownames(res_gender_age_08)[!is.na(res_gender_age_08$padj) & res_gender_age_08$padj < 0.05]

# Number of genes in common
n_common_08 <- length(intersect(signif_gender_08, signif_gender_age_08))
cat("number of overlapping significant genes when modelling with gender and age, and just gender: ",n_common_08)
```

Common significant genes between 1st run(groups) and 3rd run(groups, gender, age)

```{r}
n_sig_gender_age_08 <- sum(res_gender_age_08$padj < 0.05, na.rm = TRUE)
cat("number of significant genes when modelling with gender and age:", n_sig_gender_age_08, "\n")

cat("number of significant genes when modelling on condition:", n_sig_08, "\n")

# What is the overlap between the differential analysis with and without modeling age?
signif_08 <- rownames(res_08)[!is.na(res_08$padj) & res_08$padj < 0.05]
signif_gender_age_08 <- rownames(res_gender_age_08)[!is.na(res_gender_age_08$padj) & res_gender_age_08$padj < 0.05]

# Number of genes in common
n_common_08 <- length(intersect(signif_08, signif_gender_age_08))
cat("number of overlapping significant genes when modelling with gender and age, and just condition: ",n_common_08)
```

Age has a big influence on gene expression, but is not necessarily correlated with the condition

Conclusion: age is not a confounder!

```{r}
saveRDS(res_gender_age_08, file = "../data/deseq2_results_08_age.rds")
```

### DESeq Condition&Age

TODO!

```{r}
dds_gender_08 <- DESeqDataSetFromMatrix(
  countData = gene_count_08,
  colData = metadata_df_08,
  design = ~ gender + age)
```

```{r}
```

# Gene-set enrichment analysis

!! Wrong results so far!!

TODO: Do it with res_gender_age_08

## FSC

```{r}
BP_df = msigdbr(species = "human", category = "C5", subcategory = "BP")
BP_list = split(BP_df$ensembl_gene, BP_df$gs_name)
```

```{r}
# Create the ranked list of gene-level statistics
myStats_08 <- res_gender_08[,"stat"]
names(myStats_08) <- rownames(res_gender_08)

# Remove the ensembl version in the names
clean_names_08 <- sub("\\.\\d+$", "", names(myStats_08))
names(myStats_08) <- clean_names_08

myStats_08 <- myStats_77[!is.na(myStats_08)]
myStats_08 <- sort(myStats_08, decreasing = TRUE)
```

```{r}
fg_08 <- fgseaMultilevel(pathways = BP_list, stats = myStats_08, minSize = 15, maxSize = 500, scoreType = "std")
```

```{r}
fg_sig_08 <- fg_08 |> arrange(padj) |> filter(padj < 0.05)

nrow(fg_sig_08)
```

```{r}
fg_sig_08[1:10]
```

COMMENT ON THIS

```{r}
chosen_pathway_08 <- fg_sig_08$pathway[1]
enrichment_08_direct <- plotEnrichment(BP_list[[chosen_pathway_08]], myStats_08) + ggplot2::labs(title = paste0("Enrichment: ", chosen_pathway_08))
enrichment_08_direct
```

```{r}
ggsave("../fig/enrichment_08_direct.png", plot = enrichment_08_direct)
```

```{r}
leading_08 <- fg_sig_08$leadingEdge[[1]]
first_genes_08 <- head(leading_08, 3)
first_genes_08
```

## FSC including age

```{r}
# Create the ranked list of gene-level statistics
myStats_08 <- res_gender_age_08[,"stat"]
names(myStats_08) <- rownames(res_gender_age_08)

# Remove the ensembl version in the names
clean_names_08 <- sub("\\.\\d+$", "", names(myStats_08))
names(myStats_08) <- clean_names_08

myStats_08 <- myStats_08[!is.na(myStats_08)]
myStats_08 <- sort(myStats_08, decreasing = TRUE)
```

```{r}
fg_08 <- fgseaMultilevel(pathways = BP_list, stats = myStats_08, minSize = 15, maxSize = 500, scoreType = "std")
```

```{r}
fg_sig_08 <- fg_08 |> arrange(padj) |> filter(padj < 0.05)

nrow(fg_sig_08)
```

```{r}
fg_sig_08[1:100, c("pathway", "padj", "NES")]
```

```{r}
enrichment_08_direct <- plotEnrichment(BP_list[[chosen_pathway_08]], myStats_08)
enrichment_08_direct <- enrichment_08_direct + ggplot2::labs(title = paste0("Enrichment: ", chosen_pathway_08))
enrichment_08_direct

# save plot
ggsave("../fig/enrichment_08_direct_age.png", plot = enrichment_08_direct)
```

COMMENT ON THIS

## FSC as non-directional including age

```{r}
chosen_pathway_08 <- fg_sig_08$pathway[1]
enrichment_08_direct <- plotEnrichment(BP_list[[chosen_pathway_08]], myStats_08) +ggplot2::labs(title = paste0("Enrichment: ", chosen_pathway_08))
enrichment_08_direct
```

```{r}
stats_abs_08 <- abs(myStats_08)
fg_nondir_08 <- fgseaMultilevel(pathways = BP_list, stats = stats_abs_08,
minSize = 15, maxSize = 500, scoreType = "pos")
```

```{r}
fg_nondir_08 <- fg_nondir_08[order(fg_nondir_08$padj), ]
fg_nondir_08[1:10]
```

COMMENT ON THIS

```{r}
enrichment_08_non_direct <- plotEnrichment(BP_list[[chosen_pathway_08]], stats_abs_08)
enrichment_08_non_direct <- enrichment_08_non_direct + ggplot2::labs(title = paste0("Enrichment: ", chosen_pathway_08))

enrichment_08_non_direct
# save plot
ggsave("../fig/enrichment_08_non_direct_age.png", plot = enrichment_08_non_direct)
```

## FSC

```{r}
BP_df = msigdbr(species = "human", category = "C5", subcategory = "BP")
BP_list = split(BP_df$ensembl_gene, BP_df$gs_name)
```

```{r}
# Create the ranked list of gene-level statistics
myStats_08 <- res_gender_08[,"stat"]
names(myStats_08) <- rownames(res_gender_08)

# Remove the ensembl version in the names
clean_names_08 <- sub("\\.\\d+$", "", names(myStats_08))
names(myStats_08) <- clean_names_08

myStats_08 <- myStats_08[!is.na(myStats_08)]
myStats_08 <- sort(myStats_08, decreasing = TRUE)
```

```{r}
fg_08 <- fgseaMultilevel(pathways = BP_list, stats = myStats_08, minSize = 15, maxSize = 500, scoreType = "std")
```

```{r}
fg_sig_08 <- fg_08 |> arrange(padj) |> filter(padj < 0.05)

nrow(fg_sig_08)
```

```{r}
fg_sig_08[1:100, c("pathway", "padj", "NES")]
```

Positive regulation increases expression

Among the top 10 significant gene-set, there is a clear downregulation of immune repose, T cell activation, Leukocyte proliferation theme.

α-synuclein-specific T cells appear in PD; T cells infiltrate brain and promote neurodegeneration

In the most significant pathway, discover three genes that are implicated in this significant DE.

```{r}
chosen_pathway_08 <- fg_sig_08$pathway[1]
enrichment_08_direct <- plotEnrichment(BP_list[[chosen_pathway_08]], myStats_08) + ggplot2::labs(title = paste0("Enrichment: ", chosen_pathway_08))

enrichment_08_direct
```

```{r}
ggsave("../fig/enrichment_08_direct.png", plot = enrichment_08_direct)
```

```{r}
leading_08 <- fg_sig_08$leadingEdge[[1]]
first_genes_08 <- head(leading_08, 3)
first_genes_08
```

ENSG00000136235 (GPNMB):

ENSG00000155659 (VSIG4): Phagocytic receptor, strong negative regulator of T-cell proliferation and IL2 production. 

ENSG00000229295 (HLA-DPB1) : Important player in presenting antigen to APCs

all upregulated

## FSC as non-directional

```{r}
stats_abs_08 <- abs(myStats_08)
fg_nondir_08 <- fgseaMultilevel(pathways = BP_list, stats = stats_abs_08,
minSize = 15, maxSize = 500, scoreType = "pos")
```

```{r}
fg_nondir_08 <- fg_nondir_08[order(fg_nondir_08$padj), ]
fg_nondir_08[1:10]
```
