---
title: "DEXSeq"
format: html
---

# Setting up data

```{r}
library(DEXSeq)
library(fgsea)
library(msigdbr)
library(stringr)

load(file='/home/projects/22140/exam_data/exam_functions.Rdata')

isoform_count_08 <-loadIsoformCountMatrix('GSE106608')
isoform_count_77 <-loadIsoformCountMatrix('GSE208783,GSE208784')

gene_count_08 <- extractGeneCountMatrixFromIsoCounts(isoform_count_08)
gene_count_77 <- extractGeneCountMatrixFromIsoCounts(isoform_count_77)

isoform_count_08 <- as.data.frame(isoform_count_08)
isoform_count_77 <- as.data.frame(isoform_count_83)

isoInfo <- read_rds(
paste0(
'/home/projects/22140/exam_data/',
'Homo_sapiens.GRCh38.107.chr_patch_hapl_scaff_tx_info.Rds'
)
)
```

### Meta data

```{r}
file_path_08 <- "../metadata_temporary/Template_group6_GSE106608.csv"

metadata_df_08 <- read.csv(
  file_path_08,
  skip = 6,
  stringsAsFactors = FALSE
)
colnames(metadata_df_08)[5] <- "groups"

head(metadata_df_08)

```

```{r}
file_path_77 <- "../metadata_temporary/Template_group6_GSE128177.csv"

metadata_df_77 <- read.table(
  file_path_77,
  sep = ";",
  header = TRUE,
  skip = 6,
  stringsAsFactors = FALSE,
  quote = "\""   # Handle quoted fields
)

colnames(metadata_df_77)[5] <- "groups"

# Remove quotes from groups column
metadata_df_77$groups <- gsub("\"", "", metadata_df_77$groups)

head(metadata_df_77)
```

### Splitting characteristics into gender and age

#### 08 datafile

```{r}
metadata_df_08 <- metadata_df_08 |>
  mutate(gender = str_extract(characteristics, "(?<=Sex: )\\w$")
  )

metadata_df_08 <- metadata_df_08 |>
  mutate(
    age = as.integer(str_extract(characteristics, "(?<=age: )\\d+"))
  )

head(metadata_df_08)
```

#### 77 datafile

```{r}
metadata_df_77 <- metadata_df_77 |>
  mutate(
    gender = str_extract(characteristics, "(?<=gender: )[^,]+")
  )

metadata_df_77 <- metadata_df_77 |>
  mutate(
    age = as.integer(str_extract(characteristics, "(?<=age: )\\d+"))
  )

head(metadata_df_77)
```

```{r}
# Subset isoInfo to match the transcripts present in the count matrix
isoInfo_sub <- isoInfo[match(rownames(isoform_count_08), isoInfo$transcript_id), ]

featureID <- isoInfo_sub$transcript_id
groupID <- isoInfo_sub$gene_id

# Build DEXSeq object
dxdSubset <- DEXSeqDataSet(
  countData = isoform_count_08,
  sampleData = metadata_df_08,
  # Final successful design formula
  design = ~ sample + exon + groups:exon + gender:exon + age:exon,
  featureID = featureID,
  groupID = groupID
)

# Estimate size factors for normalization
dxdSubset <- estimateSizeFactors(dxdSubset)

# Estimate dispersions
dxdSubset <- estimateDispersions(dxdSubset)

# Test for Differential Exon Usage (DEU)
dxdSubset <- testForDEU(
  dxdSubset,
  fullModel = ~ sample + exon + condition:exon + gender:exon + age:exon,
  reducedModel = ~ sample + exon + gender:exon
)

# Estimate exon fold changes (fixing the error seen in the log)
dxdSubset <- estimateExonFoldChanges(dxdSubset, fitExpToVar="groups")
```

```{r}
dxdResult <- DEXSeqResults(dxdSubset)
alpha <- 0.05
# Significant isoforms (rows)
sig_iso <- dxdResult[!is.na(dxdResult$padj) & dxdResult$padj < alpha, ]
n_sig_isoforms <- nrow(sig_iso)
print(n_sig_isoforms)
n_sig_genes <- length(unique(sig_iso$groupID))
print(n_sig_genes)
```

```{r}
dexseq_df <- as.data.frame(dexseqResult)
dexseq_df <- dexseq_df[order(dexseq_df$padj), ]
top5_genes <- dexseq_df$groupID[!is.na(dexseq_df$padj)][1:10]
```

```{r}
# Save the complete DEXSeq results object
saveRDS(dxdResult, file = "dexseq_results.rds")

# Save significant isoforms as CSV
write.csv(
  sig_iso,
  file = "significant_isoforms.csv",
  row.names = TRUE
)

# Save ordered results data frame
write.csv(
  dexseq_df,
  file = "dexseq_results_all.csv",
  row.names = TRUE
)

# Save summary statistics as a text file
summary_stats <- data.frame(
  n_significant_isoforms = n_sig_isoforms,
  n_significant_genes = n_sig_genes,
  alpha_threshold = alpha,
  total_isoforms_tested = nrow(dxdResult),
  analysis_date = Sys.Date()
)

write.csv(
  summary_stats,
  file = "analysis_summary.csv",
  row.names = FALSE
)

# Optional: Save the complete DEXSeq object for later use
saveRDS(dxdSubset, file = "dexseq_dataset_object.rds")
```
