---
title: "Dexseqanalysis"
format: html
---

# Dexseq vs Deseq analysis

```{r}
#converting to dataframe
dxdResult_08_df <- as.data.frame(dxdResult_08)
dxdResult_77_df <- as.data.frame(dxdResult_77)

#ordering them:
dxdResult_08_df <- dxdResult_08_df[order(dxdResult_08_df$padj),]
dxdResult_77_df <- dxdResult_77_df[order(dxdResult_77_df$padj),]

```

```{r}
#making the deseq into dataframes
deseq_08_df <- as.data.frame(res_gender_age_08)
deseq_77_df <- as.data.frame(res_gender_77)
deseq_08_df$gene_id <- rownames(deseq_08_df)
deseq_77_df$gene_id <- rownames(deseq_77_df)
```

## comparing to differential expression:

#### function that turns Dexseq into gene counts

```{r}
dexseq_to_genelevel <- function(x) {
  x %>%
    as.data.frame() %>%
    as_tibble() %>%
    # Remove rows with NULL or NA 'stat'
    filter(!is.na(stat)) %>%
    # Select & rename columns
    dplyr::select(
      gene_id   = groupID,
      isoform_id = featureID,
      stat,
      log2FC = log2fold_parkinson.s_healthy,
      pvalue,
      padj
    ) %>%
    # Group isoforms by gene
    group_by(gene_id) %>%
    # Keep isoform with smallest p-value
    slice_min(order_by = pvalue, n = 1, with_ties = FALSE) %>%
    ungroup() %>%
    # Recalculate gene-level FDR
    mutate(
      padj = p.adjust(pvalue, method = "fdr")
    )
}
```

### 08 analysis

```{r}
# dexseq per-gene q-values
alpha <- 0.05
dex_gene_08 <- dexseq_to_genelevel(dxdResult_08)

dex_gene_08 <- dex_gene_08 %>%
  mutate(gene_id, dex_padj_08 = padj,
         dex_sig_08 = !is.na(padj) & padj < alpha)
#Deseq2 gene_level q-values
deseq_gene_08 <- deseq_08_df %>%
  mutate(gene_id, deseq_padj = padj,
         deseq_sig_08 = !is.na(padj) & padj < alpha)

#gene universe = genes present in both results

univ_08 <- intersect(dex_gene_08$gene_id, deseq_gene_08$gene_id)

both_08 <- dex_gene_08 %>%
  filter(gene_id %in% univ_08) %>%
  mutate(gene_id, dex_sig_08) %>%
  inner_join(deseq_gene_08 %>%
               filter(gene_id %in% univ_08) %>%
               mutate(gene_id, deseq_sig_08), by = "gene_id")
# Contigency table(dexseq vs deseq significance)
contig_tab_08 <- table(DEXseq_08 = both_08$dex_sig_08, DESeq2_08 = both_08$deseq_sig_08)
contig_tab_08

```

### fisher test for the 08 testset

```{r}
#fisherstest
ft_08 <- fisher.test(contig_tab_08, alternative = "greater")
ft_08
```

**This means that it is not significantly enriched. In other words, they describe different layers of what we are researching as there is not a big overlap.**

## comparing to differential gene expression at gene-level

```{r}
BP_df = msigdbr(species = "human", category = "C5", subcategory = "BP")
BP_list = split(BP_df$ensembl_gene, BP_df$gs_name)
```

```{r}
# Create the ranked list of gene-level statistics
deseq_names_08 <- sub("\\.\\d+$", "", deseq_gene_08$gene_id)
deseq_vals_08 <- deseq_gene_08$stat
stats_deseq_08 <- setNames(deseq_vals_08, deseq_names_08)
stats_deseq_08 <- stats_deseq_08[!is.na(stats_deseq_08)]
stats_deseq_08 >- sort(stats_deseq_08, decreasing = TRUE)
```
