---
title: "DEXSeq"
format: html
---

# Setting up data

```{r}
#| message: false
#| warning: false
#| note: false
library(DEXSeq)
library(fgsea)
library(msigdbr)
library(stringr)

load(file='/home/projects/22140/exam_data/exam_functions.Rdata')

isoform_count_08 <-loadIsoformCountMatrix('GSE106608')
isoform_count_77 <-loadIsoformCountMatrix('GSE128177')

gene_count_08 <- extractGeneCountMatrixFromIsoCounts(isoform_count_08)
gene_count_77 <- extractGeneCountMatrixFromIsoCounts(isoform_count_77)

isoform_count_08 <- as.data.frame(isoform_count_08)
isoform_count_77 <- as.data.frame(isoform_count_77)

isoInfo <- read_rds(
paste0(
'/home/projects/22140/exam_data/',
'Homo_sapiens.GRCh38.107.chr_patch_hapl_scaff_tx_info.Rds'
)
)
```

### Meta data

```{r}
file_path_08 <- "../metadata_temporary/Template_group6_GSE106608.csv"

metadata_df_08 <- read.csv(
  file_path_08,
  skip = 6,
  stringsAsFactors = FALSE
)
colnames(metadata_df_08)[5] <- "groups"

head(metadata_df_08)

```

```{r}
file_path_77 <- "../metadata_temporary/Template_group6_GSE128177.csv"

metadata_df_77 <- read.table(
  file_path_77,
  sep = ";",
  header = TRUE,
  skip = 6,
  stringsAsFactors = FALSE,
  quote = "\""   # Handle quoted fields
)

colnames(metadata_df_77)[5] <- "groups"

# Remove quotes from groups column
metadata_df_77$groups <- gsub("\"", "", metadata_df_77$groups)

head(metadata_df_77)
```

### Splitting characteristics into gender and age

#### 08 datafile

```{r}
metadata_df_08 <- metadata_df_08 |>
  mutate(gender = str_extract(characteristics, "(?<=Sex: )\\w$")
  )

metadata_df_08 <- metadata_df_08 |>
  mutate(
    age = as.integer(str_extract(characteristics, "(?<=age: )\\d+"))
  )

head(metadata_df_08)
```

#### 77 datafile

```{r}
metadata_df_77 <- metadata_df_77 |>
  mutate(
    gender = str_extract(characteristics, "(?<=gender: )[^,]+")
  )

metadata_df_77 <- metadata_df_77 |>
  mutate(
    age = as.integer(str_extract(characteristics, "(?<=age: )\\d+"))
  )

head(metadata_df_77)
```

```{r}
nrow(metadata_df_77)
```

```{r}
#| message: false
#| warning: false
filter_low_expression <- function(count_matrix, metadata = NULL, threshold = 10) {
  
  n_samples <- ifelse(is.null(metadata), ncol(count_matrix), nrow(metadata))
  min_samples <- ceiling(n_samples / 2) + 1
  high_expr_counts <- apply(count_matrix, 1, function(row) sum(row > threshold))
  keep_rows <- high_expr_counts >= min_samples
  return(count_matrix[keep_rows, , drop = FALSE])
}

# Apply filtering to both datasets
### !ATTENTIONE! we put the threshold mega high here, it shoud probably be close to 0, maybe 5. But we just do this before we figure out if we should run the entire thing, it takes a long time
isoform_count_08 <- filter_low_expression(isoform_count_08, metadata = metadata_df_08, threshold = 10)
isoform_count_77 <- filter_low_expression(isoform_count_77, metadata = metadata_df_77, threshold = 10)

# Subset isoInfo to match the transcripts present in the count matrix
## We should also do this subset info matching for count_77

isoInfo_sub <- isoInfo[match(rownames(isoform_count_08), isoInfo$transcript_id), ]

featureID <- isoInfo_sub$transcript_id
groupID <- isoInfo_sub$gene_id

# Build DEXSeq object
dxdSubset <- DEXSeqDataSet(
  countData = isoform_count_08,
  sampleData = metadata_df_08,
  # Final successful design formula
  design = ~ sample + exon + groups:exon + gender:exon,
  featureID = featureID,
  groupID = groupID
)

```

```{r}
#| message: false
#| warning: false
# Estimate size factors for normalization
dxdSubset <- estimateSizeFactors(dxdSubset)

# Estimate dispersions
dxdSubset <- estimateDispersions(dxdSubset)

# Test for Differential Exon Usage (DEU)
dxdSubset <- testForDEU(
  dxdSubset,
  fullModel = ~ sample + exon + groups:exon + gender:exon,
  reducedModel = ~ sample + exon + gender:exon
)

# Estimate exon fold changes (fixing the error seen in the log)
dxdSubset <- estimateExonFoldChanges(dxdSubset, fitExpToVar="groups")
```

```{r}
#| message: false
#| warning: false
dxdResult <- DEXSeqResults(dxdSubset)
alpha <- 0.05
# Significant isoforms (rows)
sig_iso <- dxdResult[!is.na(dxdResult$padj) & dxdResult$padj < alpha, ]
n_sig_isoforms <- nrow(sig_iso)
print(n_sig_isoforms)
n_sig_genes <- length(unique(sig_iso$groupID))
print(n_sig_genes)
```

```{r}
# See the most significant genes
dexseq_df <- as.data.frame(dxdResult)
dexseq_df <- dexseq_df[order(dexseq_df$padj), ]
top10_genes <- dexseq_df$groupID[!is.na(dexseq_df$padj)][1:10]
top10_genes
```

```{r}
# Save the complete DEXSeq results object, so we can load them again
saveRDS(dxdResult, file = "../data/dexseq_results_08.rds")
saveRDS(dxdSubset, file = "../data/dexseq_object_08.rds")
```

Now do this for dataset 77

```{r}
#| message: false
#| warning: false
isoInfo_sub_77 <- isoInfo[match(rownames(isoform_count_77), isoInfo$transcript_id), ]

featureID_77 <- isoInfo_sub_77$transcript_id
groupID_77 <- isoInfo_sub_77$gene_id

# Build DEXSeq object
dxdSubset_77 <- DEXSeqDataSet(
  countData = isoform_count_77,
  sampleData = metadata_df_77,
  # Final successful design formula
  design = ~ sample + exon + groups:exon + gender:exon,
  featureID = featureID_77,
  groupID = groupID_77
)
```

```{r}
#| message: false
#| warning: false
# Estimate size factors for normalization
dxdSubset_77 <- estimateSizeFactors(dxdSubset_77)

# Estimate dispersions
dxdSubset_77 <- estimateDispersions(dxdSubset_77)

# Test for Differential Exon Usage (DEU)
dxdSubset_77 <- testForDEU(
  dxdSubset_77,
  fullModel = ~ sample + exon + groups:exon + gender:exon,
  reducedModel = ~ sample + exon + gender:exon
)

# Estimate exon fold changes (fixing the error seen in the log)
dxdSubset_77 <- estimateExonFoldChanges(dxdSubset_77, fitExpToVar="groups")
```

```{r}
#| message: false
#| warning: false
dxdResult_77 <- DEXSeqResults(dxdSubset_77)
alpha <- 0.05
# Significant isoforms (rows)
sig_iso <- dxdResult_77[!is.na(dxdResult_77$padj) & dxdResult_77$padj < alpha, ]
n_sig_isoforms <- nrow(sig_iso)
print(n_sig_isoforms)
n_sig_genes <- length(unique(sig_iso$groupID))
print(n_sig_genes)
```

```{r}
# See the most significant genes
dexseq_df <- as.data.frame(dxdResult_77)
dexseq_df <- dexseq_df[order(dexseq_df$padj), ]
top10_genes <- dexseq_df$groupID[!is.na(dexseq_df$padj)][1:10]
top10_genes
```

```{r}
saveRDS(dxdResult_77, file = "../data/dexseq_results_77.rds")
saveRDS(dxdSubset_77, file = "../data/dexseq_object_77.rds")
```

```{r}
test <- readRDS("../data/dexseq_results_77_full.rds")
test
```
