---
title: "DEXSeq"
format: html
---

# Filtering of low expression Genes

```{r}
#| message: false
#| warning: false
filter_low_expression <- function(count_matrix, metadata = NULL, threshold = 10) {
  
  n_samples <- ifelse(is.null(metadata), ncol(count_matrix), nrow(metadata))
  min_samples <- ceiling(n_samples / 2) + 1
  high_expr_counts <- apply(count_matrix, 1, function(row) sum(row > threshold))
  keep_rows <- high_expr_counts >= min_samples
  return(count_matrix[keep_rows, , drop = FALSE])
}

```

```{r}
# Apply filtering to both datasets
### !ATTENTIONE! we put the threshold mega high here, it shoud probably be close to 0, maybe 5. But we just do this before we figure out if we should run the entire thing, it takes a long time
isoform_count_08 <- filter_low_expression(isoform_count_08, metadata = metadata_df_08, threshold = 10)
isoform_count_77 <- filter_low_expression(isoform_count_77, metadata = metadata_df_77, threshold = 10)


dim(isoform_count_08)
dim(isoform_count_77)
```

```{r}
# Subset isoInfo to match the transcripts present in the count matrix
## We should also do this subset info matching for count_77

isoInfo_sub <- isoInfo[match(rownames(isoform_count_08), isoInfo$transcript_id), ]

featureID <- isoInfo_sub$transcript_id
groupID <- isoInfo_sub$gene_id

```

# DEXSeq 08: Condition&Gender

```{r}
# Build DEXSeq object
dxdSubset_08 <- DEXSeqDataSet(
  countData = isoform_count_08,
  sampleData = metadata_df_08,
  # Final successful design formula
  design = ~ sample + exon + groups:exon + gender:exon,
  featureID = featureID,
  groupID = groupID
)
```

```{r}
#| message: false
#| warning: false
# Estimate size factors for normalization
dxdSubset_08 <- estimateSizeFactors(dxdSubset_08)

# Estimate dispersions
dxdSubset_08 <- estimateDispersions(dxdSubset_08)


# Test for Differential Exon Usage (DEU)
dxdSubset_08 <- testForDEU(
  dxdSubset_08,
  fullModel = ~ sample + exon + groups:exon + gender:exon,
  reducedModel = ~ sample + exon + gender:exon
)

# Estimate exon fold changes (fixing the error seen in the log)
dxdSubset_08 <- estimateExonFoldChanges(dxdSubset_08, fitExpToVar="groups")
```

```{r}
#| message: false
#| warning: false
dxdResult_08 <- DEXSeqResults(dxdSubset_08)
alpha <- 0.05
#dxdResult <- dxdResult_08
# Significant isoforms (rows)
sig_iso <- dxdResult_08[!is.na(dxdResult_08$padj) & dxdResult_08$padj < alpha, ]
n_sig_isoforms <- nrow(sig_iso)
print(n_sig_isoforms)
n_sig_genes <- length(unique(sig_iso$groupID))
print(n_sig_genes)
```

```{r}
# See the most significant genes
dexseq_df_08 <- as.data.frame(dxdResult_08)
dexseq_df_08 <- dexseq_df_08[order(dexseq_df_08$padj), ]
top10_genes <- dexseq_df_08$groupID[!is.na(dexseq_df_08$padj)][1:10]
top10_genes
```

```{r}
# Save the complete DEXSeq results object, so we can load them again
saveRDS(dxdResult_08, file = "../data/dexseq_results_08.rds")
```

# DEXSeq 77: Condition&Gender

Now do this for dataset 77

```{r}
#| message: false
#| warning: false
isoInfo_77 <- isoInfo[match(rownames(isoform_count_77), isoInfo$transcript_id), ]

featureID_77 <- isoInfo_77$transcript_id
groupID_77 <- isoInfo_77$gene_id

# Build DEXSeq object
dxdSubset_77 <- DEXSeqDataSet(
  countData = isoform_count_77,
  sampleData = metadata_df_77,
  # Final successful design formula
  design = ~ sample + exon + groups:exon + gender:exon,
  featureID = featureID_77,
  groupID = groupID_77
)
```

```{r}
#| message: false
#| warning: false
# Estimate size factors for normalization
dxdSubset_77 <- estimateSizeFactors(dxdSubset_77)

# Estimate dispersions
dxdSubset_77 <- estimateDispersions(dxdSubset_77)

# Test for Differential Exon Usage (DEU)
dxdSubset_77 <- testForDEU(
  dxdSubset_77,
  fullModel = ~ sample + exon + groups:exon + gender:exon,
  reducedModel = ~ sample + exon + gender:exon
)

# Estimate exon fold changes (fixing the error seen in the log)
dxdSubset_77 <- estimateExonFoldChanges(dxdSubset_77, fitExpToVar="groups")
```

```{r}
#| message: false
#| warning: false
dxdResult_77 <- DEXSeqResults(dxdSubset_77)
alpha <- 0.05
# Significant isoforms (rows)
sig_iso <- dxdResult_77[!is.na(dxdResult_77$padj) & dxdResult_77$padj < alpha, ]
n_sig_isoforms <- nrow(sig_iso)
print(n_sig_isoforms)
n_sig_genes <- length(unique(sig_iso$groupID))
print(n_sig_genes)
```

```{r}
# See the most significant genes
dexseq_df_77 <- as.data.frame(dxdResult_77)
dexseq_df_77 %>% 
  na.omit() %>% 
  filter(padj < 0.05)

dexseq_df_77 <- dexseq_df_77[order(dexseq_df_77$padj), ]
top10_genes <- dexseq_df_77$groupID[!is.na(dexseq_df_77$padj)][1:10]
top10_genes
```

```{r}
saveRDS(dxdResult_77, file = "../data/dexseq_results_77.rds")
```

# DEXSeq Subset77: Condition&Gender

```{r}
#| message: false
#| warning: false
isoInfo_sub_77 <- isoInfo[match(rownames(isoform_count_77_sub), isoInfo$transcript_id), ]

featureID_77 <- isoInfo_sub_77$transcript_id
groupID_77 <- isoInfo_sub_77$gene_id

# Build DEXSeq object
dxdSubset_77_sub <- DEXSeqDataSet(
  countData = isoform_count_77_sub,
  sampleData = metadata_df_77_sub,
  # Final successful design formula
  design = ~ sample + exon + groups:exon + gender:exon,
  featureID = featureID_77,
  groupID = groupID_77
)
```

```{r}
#| message: false
#| warning: false
# Estimate size factors for normalization
dxdSubset_77_sub <- estimateSizeFactors(dxdSubset_77_sub)

# Estimate dispersions
dxdSubset_77_sub <- estimateDispersions(dxdSubset_77_sub)

# Test for Differential Exon Usage (DEU)
dxdSubset_77_sub <- testForDEU(
  dxdSubset_77_sub,
  fullModel = ~ sample + exon + groups:exon + gender:exon,
  reducedModel = ~ sample + exon + gender:exon
)

# Estimate exon fold changes (fixing the error seen in the log)
dxdSubset_77_sub <- estimateExonFoldChanges(dxdSubset_77_sub, fitExpToVar="groups")
```

```{r}
#| message: false
#| warning: false
dxdResult_77_sub <- DEXSeqResults(dxdSubset_77_sub)
alpha <- 0.05
# Significant isoforms (rows)
sig_iso <- dxdResult_77_sub[!is.na(dxdResult_77_sub$padj) & dxdResult_77_sub$padj < alpha, ]
n_sig_isoforms <- nrow(sig_iso)
print(n_sig_isoforms)
n_sig_genes <- length(unique(sig_iso$groupID))
print(n_sig_genes)
```

```{r}
# See the most significant genes
dexseq_df_77_sub <- as.data.frame(dxdResult_77_sub)
dexseq_df_77_sub %>% 
  na.omit() %>% 
  filter(padj < 0.05)

dexseq_df_77_sub <- dexseq_df_77_sub[order(dexseq_df_77_sub$padj), ]
top10_genes <- dexseq_df_77_sub$groupID[!is.na(dexseq_df_77_sub$padj)][1:10]
top10_genes
```

```{r}
saveRDS(dxdResult_77_sub, file = "../data/dexseq_results_77_sub.rds")
```
