---
title: "exploratory_analysis"
format: html
---

# Load Libraries

```{r}
library(DEXSeq)
library(fgsea)
library(msigdbr)
library(tidyverse)
library(matrixStats)
```

# Setting up data

```{r}
load(file='/home/projects/22140/exam_data/exam_functions.Rdata')

isoform_count_08 <-loadIsoformCountMatrix('GSE106608')
isoform_count_83 <-loadIsoformCountMatrix('GSE208783,GSE208784')

gene_count_08 <- extractGeneCountMatrixFromIsoCounts(isoform_count_08)
gene_count_83 <- extractGeneCountMatrixFromIsoCounts(isoform_count_83)

isoform_count_08 <- as.data.frame(isoform_count_08)
isoform_count_83 <- as.data.frame(isoform_count_83)

isoInfo <- read_rds(
paste0(
'/home/projects/22140/exam_data/',
'Homo_sapiens.GRCh38.107.chr_patch_hapl_scaff_tx_info.Rds'
)
)
```

### Meta data

```{r}
file_path_08 <- "../metadata_temporary/Template_group6_GSE106608.csv"

metadata_df_08 <- read.csv(
  file_path_08,
  skip = 6,
  stringsAsFactors = FALSE
)
colnames(metadata_df_08)[5] <- "groups"
head(metadata_df_08)
```

```{r}
file_path_83 <- "../metadata_temporary/Template_group6_GSE208783.csv"
metadata_df_83 <- read.table(
  file_path_83,
  sep = ",",          
  header = TRUE,
  skip = 6,
  stringsAsFactors = FALSE,
  quote = "\""         # Add this line to handle quoted fields
)
colnames(metadata_df_83)[5] <- "groups"
# Remove the "" from groups
metadata_df_83$groups <- gsub("\"", "", metadata_df_83$groups)

head(metadata_df_83)
```

# Exploratory Analysis

## Genes

How many genes

```{r}
dim(gene_count_08)
dim(gene_count_83)
```

empty rows? Not expressed genes overlapping?

```{r}
gene_count_08 %>% 
  filter(rowSums(as.matrix(.))==0) %>% 
  nrow()

gene_count_83 %>% 
  filter(rowSums(as.matrix(.))==0) %>% 
  nrow()

zero_expression_08 <- gene_count_08 %>% 
  filter(rowSums(as.matrix(.))==0) %>% 
  rownames()

zero_expression_83 <- gene_count_83 %>% 
  filter(rowSums(as.matrix(.))==0) %>% 
  rownames()

zero_expression_both <- intersect(zero_expression_08, zero_expression_83)

# clean from empty rows
complete_gene_08 <- gene_count_08 %>% 
  filter(rowSums(as.matrix(.))!=0) 
complete_gene_83 <- gene_count_83 %>% 
  filter(rowSums(as.matrix(.))!=0)
```

Statistics of expression values

```{r}
gene_stats <- complete_gene_08 %>%
  mutate(
    min = rowMins(as.matrix(.)),
    max  = rowMaxs(as.matrix(.)),
    mean = rowMeans(as.matrix(.)),
    sd   = rowSds(as.matrix(.)),
    sum = rowSums2(as.matrix(.))
  ) %>%
  ungroup()
```

```{r}
ggplot(gene_stats, aes(x = mean)) +
  geom_histogram(bins = 30, fill = "skyblue", color = "black") +
  labs(title = "Distribution of Mean Gene Expression",
       x = "Mean Expression Value",
       y = "Number of Genes") +
  theme_minimal()

gene_stats %>% select(min, max, mean, sd) %>%   summary()
```

How many protein coding genes,...

```{r}
geneInfo <- isoInfo %>% 
  select(gene_id, gene_name, gene_biotype) %>% 
  distinct()

gene_info_08 <- complete_gene_08 %>% 
  mutate(gene_id = row.names(.)) %>% 
  left_join(geneInfo, by = 'gene_id')

gene_info_08 %>% filter(gene_biotype == 'protein_coding') %>% nrow()
unique(gene_info_08$gene_biotype)
```

## Transcripts

How many transcript?

How many empty rows?

Statistics of existence

How many protein coding genes,...

## Samples

```{r}
samples_08 <- as.data.frame(t(complete_gene_08)) %>% 
  mutate(
    sample_id = row.names(.)
  ) %>% 
  left_join(metadata_df_08, by = 'sample_id') %>% 
    select(-c(study_id, title))
dim(samples_08)

samples_08 %>% filter(is.na(groups)) %>%  nrow() #check join worked
samples_08 %>%  select(groups) %>% 
  table()
```

Statistics over samples(look for bias)
