---
title: "p_value_integration"
format: html
---

# Load libraries

```{r}
library(tidyverse)
library(fgsea)
library(msigdbr)
```

```{r}
set.seed(42)
```

# Data Loading

```{r}
res_gender_age_08 <- readRDS("../data/deseq2_results_08_age.rds")
dxdResult_08 <- readRDS("../data/dexseq_results_08_age.rds")
dxdResult_77 <- readRDS("../data/dexseq_results_77_age_sub.rds")
res_gender_age_77 <- readRDS("../data/deseq2_results_77_age.rds")
```

```{r}
diff_gene_expr_08 <- as.data.frame(res_gender_age_08) %>% 
  dplyr::select(pvalue) %>% 
  rownames_to_column(var = 'Genes') %>% 
  dplyr::rename(deseq_08_pval = pvalue )
diff_gene_expr_77 <- as.data.frame(res_gender_age_77) %>% 
  dplyr::select(pvalue) %>% 
  rownames_to_column(var = 'Genes') %>% 
  dplyr::rename(deseq_77_pval = pvalue)
diff_transcript_expr_08 <- as.data.frame(dxdResult_08) %>% 
  dplyr::select(groupID, pvalue) %>% 
  dplyr::rename(
    Genes = groupID
  ) %>% 
  group_by(Genes) %>% 
  summarise(dexseq_08_pval = min(pvalue))
diff_transcript_expr_77 <- as.data.frame(dxdResult_77) %>% 
  dplyr::select(groupID, pvalue) %>% 
  dplyr::rename(
    Genes = groupID
  ) %>% 
  group_by(Genes) %>% 
  summarise(dexseq_77_pval = min(pvalue))

dim(diff_gene_expr_08)
dim(diff_transcript_expr_08)
dim(diff_transcript_expr_77)
dim(diff_gene_expr_77)
```

# (Joining all Datasets at once)

```{r}
full_diff_analysis <- diff_gene_expr_08 %>% 
  inner_join(diff_gene_expr_77, by = 'Genes') %>% 
  inner_join(diff_transcript_expr_08, by = 'Genes') %>% 
  inner_join(diff_transcript_expr_77, by = 'Genes') %>% 
  na.omit()
full_diff_analysis
```

```{r}
full_diff_analysis <- full_diff_analysis |>
  rowwise() |>
  mutate(
  joint_pval_fisher = metap::sumlog(c(deseq_08_pval, deseq_77_pval, dexseq_08_pval, dexseq_77_pval))$p  ) |>
  ungroup()
```

## Multiple Testing adjustment

```{r}
full_diff_analysis <- full_diff_analysis %>% 
  mutate(
  padj_deseq_08 = p.adjust(deseq_08_pval, method = "BH"),
  padj_deseq_77 = p.adjust(deseq_77_pval, method = "BH"),
  padj_dexseq_08 = p.adjust(dexseq_08_pval, method = "BH"),
  padj_dexseq_77 = p.adjust(dexseq_77_pval, method = "BH"),
  padj_joint = p.adjust(joint_pval_fisher, method = "BH"),
  sig_deseq_08 = padj_deseq_08 < 0.05,
  sig_deseq_77 = padj_deseq_77 < 0.05,
  sig_dexseq_08 = padj_dexseq_08 < 0.05,
  sig_dexseq_77 = padj_dexseq_77 < 0.05,
  sig_joint = padj_joint < 0.05
  )
```

```{r}
sig_counts <- c(
DESeq_08 = sum(full_diff_analysis$sig_deseq_08),
DESeq_77 = sum(full_diff_analysis$sig_deseq_77),
DEXSeq_08 = sum(full_diff_analysis$sig_dexseq_08),
DEXSeq_77 = sum(full_diff_analysis$sig_dexseq_77),
Joint = sum(full_diff_analysis$sig_joint)
)
sig_counts
```

uniquely significant in joint

```{r}
unique_joint <- sum(full_diff_analysis$sig_joint &
!full_diff_analysis$sig_deseq_08 & !full_diff_analysis$sig_deseq_77 & !full_diff_analysis$sig_dexseq_08 & !full_diff_analysis$sig_dexseq_77)
unique_joint

```

# (Integration within Levels)

## Transcripts(Edingtons)

### Gene Counts

```{r}
gene_diff_analysis <- diff_gene_expr_08 %>% 
  inner_join(diff_gene_expr_77, by = 'Genes') %>% 
  na.omit()
dim(gene_diff_analysis)
```

```{r}
gene_diff_analysis <- gene_diff_analysis |>
  rowwise() |>
  mutate(
  joint_pval_fisher_gene = metap::sump(c(deseq_08_pval, deseq_77_pval))$p ,
  adj_pval = p.adjust(joint_pval_fisher_gene)) |>
  ungroup()
```

### Isoform Counts

```{r}
isoform_diff_analysis <-diff_transcript_expr_08 %>% 
  inner_join(diff_transcript_expr_77, by = 'Genes') %>% 
  na.omit()
dim(isoform_diff_analysis)
```

```{r}
isoform_diff_analysis <- isoform_diff_analysis |>
  rowwise() |>
  mutate(
  joint_pval_fisher_iso = metap::sump(c( dexseq_08_pval, dexseq_77_pval))$p  ) |>
  ungroup()
```

## Joining Isoform and Gene counts(fishers)

```{r}
full_diff_analysis_2 <- isoform_diff_analysis %>% 
  dplyr::select(Genes, joint_pval_fisher_iso) %>% 
  dplyr::inner_join(gene_diff_analysis %>% dplyr::select(Genes, joint_pval_fisher_gene), by='Genes')
  
full_diff_analysis_2 <- full_diff_analysis_2 |>
  rowwise() |>
  mutate(
  joint_pval_fisher = metap::sumlog(c( joint_pval_fisher_iso, joint_pval_fisher_gene))$p,
  adj_pval = p.adjust(joint_pval_fisher)
  ) |>
  ungroup()
```

```{r}
full_diff_analysis_2 %>% filter(adj_pval <0.05) %>% nrow()
```

# Joining Datasets first

## Within Datasets (Fishers)

### 08

```{r}
diff_analysis_08 <- diff_gene_expr_08 %>% 
  inner_join(diff_transcript_expr_08, by = 'Genes') %>% 
  na.omit()
dim(diff_analysis_08)
```

```{r}
int_diff_analysis_08 <- diff_analysis_08 |>
  rowwise() |>
  mutate(
  joint_pval_fisher_08 = metap::sumlog(c( deseq_08_pval, dexseq_08_pval))$p,
  adj_joint_pval = p.adjust(joint_pval_fisher_08)) |>
  ungroup()
```

```{r}
int_diff_analysis_08 %>% 
  filter(adj_joint_pval<0.05) %>% nrow()
```

### 77

```{r}
diff_analysis_77 <-diff_gene_expr_77 %>% 
  inner_join(diff_transcript_expr_77, by = 'Genes') %>% 
  na.omit()
dim(diff_analysis_77)
```

```{r}
int_diff_analysis_77 <- diff_analysis_77 |>
  rowwise() |>
  mutate(
  joint_pval_fisher_77 = metap::sumlog(c( deseq_77_pval, dexseq_77_pval))$p,
  adj_joint_pval = p.adjust(joint_pval_fisher_77)) |>
  ungroup()
```

```{r}
int_diff_analysis_77 %>% filter(adj_joint_pval <0.05) %>% nrow()
```

## Across Datasets(edingtons)

```{r}
full_diff_analysis <- int_diff_analysis_08 %>% 
  dplyr::select(Genes, joint_pval_fisher_08) %>% 
  dplyr::inner_join(int_diff_analysis_77 %>% dplyr::select(Genes, joint_pval_fisher_77), by='Genes')
  
```

```{r}
full_diff_analysis <- full_diff_analysis %>%
  na.omit() %>% 
  rowwise() %>%
  mutate(
  joint_pval_edingtons = metap::sump(c(joint_pval_fisher_08, joint_pval_fisher_77))$p,
  adj_joint_pval = p.adjust(joint_pval_edingtons, method = 'BH')
  ) |>
  ungroup()

# full_diff_analysis <- full_diff_analysis |>
#   rowwise() |>
#   dplyr::mutate(
#   joint_pval_edingtons = metap::sump(c( joint_pval_fisher_08, joint_pval_fisher_77))$p  ) |>
#   ungroup()
```

```{r}
full_diff_analysis %>% filter(adj_joint_pval<0.05) %>% nrow()
```

## Save Object for further Analysis

```{r}
int_diff_analysis_77_joint <- int_diff_analysis_77 %>% dplyr::select(Genes, adj_joint_pval)
int_diff_analysis_08_joint <- int_diff_analysis_08 %>% dplyr::select(Genes, adj_joint_pval)

saveRDS(int_diff_analysis_77_joint, file = "../data/integrated_77.rds")
saveRDS(int_diff_analysis_08_joint, file = "../data/integrated_08.rds")
```

# GSEA(full)

```{r}
library(fgsea)
library(msigdbr)
```

```{r}
BP_df = msigdbr(species = "human", category = "C5", subcategory = "BP")
BP_list = split(BP_df$ensembl_gene, BP_df$gs_name)
```

```{r}
int_stats <- full_diff_analysis$adj_joint_pval
names(int_stats) <- full_diff_analysis$Genes

int_stats <- int_stats[!is.na(int_stats)]
int_stats <- sort(int_stats, decreasing = TRUE)

fg_integrated <- fgseaMultilevel(pathways = BP_list, stats = int_stats, minSize = 15, maxSize = 500, scoreType = "std")

fg_integrated |> arrange(padj) |> filter(padj < 0.05) %>% nrow()
fg_integrated |> arrange(pval) |> filter(pval < 0.05) %>% head(n = 10)

```

# GSEA(08)

```{r}
int_stats_08 <- int_diff_analysis_08$adj_joint_pval
names(int_stats_08) <- int_diff_analysis_08$Genes

int_stats_08 <- int_stats_08[!is.na(int_stats_08)]
int_stats_08 <- sort(int_stats_08, decreasing = TRUE)

fg_integrated_08 <- fgseaMultilevel(pathways = BP_list, stats = int_stats_08, minSize = 15, maxSize = 500, scoreType = "std")

fg_integrated_08 |> arrange(padj) |> filter(padj < 0.05) %>% nrow()
fg_integrated_08 |> arrange(pval) |> filter(pval < 0.05) %>% head(n=10)
```

# GSEA 77

```{r}
int_stats_77 <- int_diff_analysis_77$adj_joint_pval
names(int_stats_77) <- int_diff_analysis_77$Genes

int_stats_77 <- int_stats_77[!is.na(int_stats_77)]
int_stats_77 <- sort(int_stats_77, decreasing = TRUE)

fg_integrated_77 <- fgseaMultilevel(pathways = BP_list, stats = int_stats_77, minSize = 15, maxSize = 500, scoreType = "std")

fg_integrated_77 |> arrange(padj) |> filter(padj < 0.05) %>% nrow()
fg_integrated_77 |> arrange(pval) |> filter(pval < 0.05) %>% head(n=10)
```

```{r}
head_77<- fg_integrated_77 |> arrange(pval) |> filter(pval < 0.05) %>% head(n=100)
head_08<- fg_integrated_08 |> arrange(pval) |> filter(pval < 0.05) %>% head(n=100)
common_pathways <- intersect(head_08$pathway, head_77$pathway)
common_pathways
```

# GSEA2

```{r}
full_diff_analysis_2

int_stats_2 <- full_diff_analysis_2$adj_pval
names(int_stats_2) <- full_diff_analysis_2$Genes

int_stats_2 <- int_stats_2[!is.na(int_stats_2)]
int_stats_2 <- sort(int_stats_2, decreasing = TRUE)

fg_integrated_2 <- fgseaMultilevel(pathways = BP_list, stats = int_stats_2, minSize = 15, maxSize = 500, scoreType = "std")

fg_integrated_2 |> arrange(padj) |> filter(padj < 0.05) %>% nrow()
fg_integrated_2 |> arrange(pval) |> filter(pval < 0.05) %>% head(n=10)
```

## per gene

```{r}
int_stats_2_gene <- gene_diff_analysis$adj_pval
names(int_stats_2_gene) <- gene_diff_analysis$Genes

int_stats_2_gene <- int_stats_2_gene[!is.na(int_stats_2_gene)]
int_stats_2_gene <- sort(int_stats_2_gene, decreasing = TRUE)

fg_integrated_2_gene <- fgseaMultilevel(pathways = BP_list, stats = int_stats_2_gene, minSize = 15, maxSize = 500, scoreType = "std")

fg_integrated_2_gene |> arrange(padj) |> filter(padj < 0.05) %>% nrow()
fg_integrated_2_gene |> arrange(pval) |> filter(pval < 0.05) %>% head(n=10)
```
